core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1206,6 +1244,41 @@ private void deleteInternal(LockedInodePath inodePath, boolean replayed, long op
     Metrics.PATHS_DELETED.inc(delInodes.size());
   }
 
+   /**
+   * Check if immediate children of directory are in sync with UFS.
+   *
+   * @param inode directory to check
+   * @param path of directory to to check
+   * @return true is contents of directory match
+   */
+  private boolean isUFSDeleteSafe(Inode inode, AlluxioURI alluxioUri)
+      throws FileDoesNotExistException, InvalidPathException, IOException {
+    Preconditions.checkArgument(inode.isDirectory());
+    Preconditions.checkArgument(inode.isPersisted());
+
+    MountTable.Resolution resolution = mMountTable.resolve(alluxioUri);
+    String ufsUri = resolution.getUri().toString();
+    UnderFileSystem ufs = resolution.getUfs();
+    UnderFileStatus[] ufsChildren =
+        ufs.listStatus(ufsUri, ListOptions.defaults().setRecursive(false));
+    // Empty directories are not persisted
+    if (ufsChildren == null) {
+      return true;
+    }
+    for (UnderFileStatus child : ufsChildren) {
+      boolean found = false;
+      for (Inode<?> inodeChild : ((InodeDirectory) inode).getChildren()) {
+        if (child.getName().equals(inodeChild.getName())) {
+          found = true;
+        }
+      }
+      if (!found) {
+        return false;
+      }
+    }
```

Comment:
The `Inode` comparator seems to compare `id` and not the `name`. How would that work?